### ISSUE TITLE
When use QUIC memory usage keep growing without decrease in 2.8.3

### ISSUE THREAD

**Issue by chnchi** [haproxy/haproxy#2357]
### Detailed Description of the Problem

When we use the QUIC protocol for access, the memory usage of HAProxy continues to grow. There is a noticeable increase in memory consumption whenever there is an increase in traffic, and the memory usage does not decrease after the traffic subsides. In the same business scenario, when the QUIC protocol is not used, the memory does not continue to grow.

1. The monitoring charts for memory usage
<img width="714" alt="image" src="https://github.com/haproxy/haproxy/assets/30403780/9fa71fd0-7ede-4bc7-8e58-6fb6c9010148">

2. Network traffic
<img width="717" alt="image" src="https://github.com/haproxy/haproxy/assets/30403780/262aae6a-511a-4dc9-ade6-5dcf633d0d07">

3. Tcp connections
<img width="715" alt="image" src="https://github.com/haproxy/haproxy/assets/30403780/683d42fb-d0f8-48c3-b11d-d8615812a6c7">

4. UDP traffic
<img width="720" alt="image" src="https://github.com/haproxy/haproxy/assets/30403780/6ffc956a-8e9a-4ed2-a1b1-61bdd2bd462d">

**In the same business scenario, when only using traditional HTTP and not adopting the QUIC protocol for access, memory usage remains stable.**
<img width="727" alt="image" src="https://github.com/haproxy/haproxy/assets/30403780/4edabd9a-76c7-4a73-a213-0b33a387278c">

 

### Expected Behavior

Memory usage should remain stable within a certain range, and even if memory usage increases with traffic growth, it should decrease after traffic subsides.

### Steps to Reproduce the Behavior

Use the QUIC protocol for API access

### What is your configuration?

```haproxy
global
  maxconn  200000
  pidfile  /var/run/haproxy.pid
  stats socket /var/run/haproxy/stats
  user  haproxy
  ca-base  /etc/pki/tls/certs/
  log 127.0.0.1:514 local6 info

defaults
  mode  http
  retries                 1
  timeout check           3s
  timeout client          600s
  timeout client-fin      5s
  timeout server          600s
  timeout server-fin      5s
  timeout connect         10s
  timeout queue           60s
  timeout tunnel          1h
  log                     global
  option                  httpslog

listen stats
  mode http
  bind 10.1.1.1:8080
  stats uri /stats
  stats refresh 10s
  stats hide-version
  http-request use-service prometheus-exporter if { path /metrics }

frontend fe
  mode http
  option forwardfor
  bind :80
  bind :443 ssl crt /etc/haproxy/certs/ssl.crt alpn h2
  bind quic4@:443 ssl crt /etc/haproxy/certs/ssl.crt alpn h3
  http-request redirect scheme https unless { ssl_fc }
  http-after-response add-header alt-svc 'h3=":443"; ma=2592000'
  default_backend nginx

backend nginx
  mode http
  balance roundrobin
  default-server port 80 inter 1s fall 3 rise 3
  server nginx-1 10.1.1.2:80 maxconn 50000 maxqueue 100 check
  server nginx-2 10.1.1.3:80 maxconn 50000 maxqueue 100 check
```

### Output of `haproxy -vv`

```plain
HAProxy version 2.8.3-86e043a 2023/09/07 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2028.
Known bugs: http://www.haproxy.org/bugs/bugs-2.8.3.html
Running on: Linux 4.19.188-10.el7.ucloud.x86_64 #1 SMP Wed Apr 28 09:54:38 UTC 2021 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wundef -Wdeclaration-after-statement -Wfatal-errors -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wno-string-plus-int -Wno-atomic-alignment
  OPTIONS = USE_OPENSSL=1 USE_ZLIB=1 USE_SYSTEMD=1 USE_QUIC=1 USE_PROMEX=1 USE_PCRE=1
  DEBUG   = -DDEBUG_STRICT -DDEBUG_MEMORY_POOLS

Feature list : -51DEGREES +ACCEPT4 +BACKTRACE -CLOSEFROM +CPU_AFFINITY +CRYPT_H -DEVICEATLAS +DL -ENGINE +EPOLL -EVPORTS +GETADDRINFO -KQUEUE -LIBATOMIC +LIBCRYPT +LINUX_CAP +LINUX_SPLICE +LINUX_TPROXY -LUA -MATH -MEMORY_PROFILING +NETFILTER +NS -OBSOLETE_LINKER +OPENSSL -OPENSSL_WOLFSSL -OT +PCRE -PCRE2 -PCRE2_JIT -PCRE_JIT +POLL +PRCTL -PROCCTL +PROMEX -PTHREAD_EMULATION +QUIC +RT +SHM_OPEN -SLZ +SSL -STATIC_PCRE -STATIC_PCRE2 +SYSTEMD +TFO +THREAD +THREAD_DUMP +TPROXY -WURFL +ZLIB

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_TGROUPS=16, MAX_THREADS=256, default=8).
Built with OpenSSL version : OpenSSL 3.1.4+quic 24 Oct 2023
Running on OpenSSL version : OpenSSL 3.1.4+quic 24 Oct 2023
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
OpenSSL providers loaded : default
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with zlib version : 1.2.7
Running on zlib version : 1.2.7
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE version : 8.32 2012-11-30
Running on PCRE version : 8.32 2012-11-30
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Encrypted password support via crypt(3): yes
Built with gcc compiler version 11.2.1 20220127 (Red Hat 11.2.1-9)

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
       quic : mode=HTTP  side=FE     mux=QUIC  flags=HTX|NO_UPG|FRAMED
         h2 : mode=HTTP  side=FE|BE  mux=H2    flags=HTX|HOL_RISK|NO_UPG
       fcgi : mode=HTTP  side=BE     mux=FCGI  flags=HTX|HOL_RISK|NO_UPG
  <default> : mode=HTTP  side=FE|BE  mux=H1    flags=HTX
         h1 : mode=HTTP  side=FE|BE  mux=H1    flags=HTX|NO_UPG
  <default> : mode=TCP   side=FE|BE  mux=PASS  flags=
       none : mode=TCP   side=FE|BE  mux=PASS  flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[BWLIM] bwlim-in
	[BWLIM] bwlim-out
	[CACHE] cache
	[COMP] compression
	[FCGI] fcgi-app
	[SPOE] spoe
	[TRACE] trace
```

### Last Outputs and Backtraces

```plain
27584> show pools
Dumping pools usage. Use SIGQUIT to flush them.
  - Pool quic_tls_iv (24 bytes) : 85660 allocated (2055840 bytes), 83967 used (~127 by thread caches), needed_avg 85216, 0 failures, 6 users, @0x2722a40 [SHARED]
  - Pool quic_arng (56 bytes) : 307628 allocated (17227168 bytes), 303236 used (~198 by thread caches), needed_avg 305569, 0 failures, 22 users, @0x26fa240 [SHARED]
  - Pool quic_cstrea (88 bytes) : 29032 allocated (2554816 bytes), 28558 used (~68 by thread caches), needed_avg 28958, 0 failures, 10 users, @0x27023c0 [SHARED]
  - Pool h3c (120 bytes) : 9121 allocated (1094520 bytes), 8882 used (~23 by thread caches), needed_avg 9045, 0 failures, 3 users, @0x271e980 [SHARED]
  - Pool ssl_sock_ct (152 bytes) : 12631830 allocated (1920038160 bytes), 12631830 used (~188 by thread caches), needed_avg 2094788, 0 failures, 9 users, @0x26f6180 [SHARED]
  - Pool quic_frame (184 bytes) : 326230 allocated (60026320 bytes), 325702 used (~269 by thread caches), needed_avg 326058, 0 failures, 7 users, @0x26fe300 [SHARED]
  - Pool tcpcheck_ru (216 bytes) : 159 allocated (34344 bytes), 38 used (~38 by thread caches), needed_avg 138, 0 failures, 4 users, @0x2732d80 [SHARED]
  - Pool ssl_sock_cl (280 bytes) : 1 allocated (280 bytes), 1 used (~0 by thread caches), needed_avg 0, 0 failures, 4 users, @0x26f2080 [SHARED]
  - Pool quic_dgram (344 bytes) : 36095 allocated (12416680 bytes), 35051 used (~100 by thread caches), needed_avg 35892, 0 failures, 2 users, @0x270e640 [SHARED]
  - Pool quic_rx_pac (376 bytes) : 8078372 allocated (3037467872 bytes), 8078372 used (~56 by thread caches), needed_avg 3998275, 0 failures, 2 users, @0x270a580 [SHARED]
  - Pool resolv_reso (472 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x273af40 [SHARED]
  - Pool resolv_answ (600 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x273f000 [SHARED]
  - Pool ehdl_sub_ed (792 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x27430c0 [SHARED]
  - Pool stream (888 bytes) : 160 allocated (142080 bytes), 42 used (~41 by thread caches), needed_avg 140, 0 failures, 1 users, @0x2736e80 [SHARED]
  - Pool quic_crypto (1048 bytes) : 66405 allocated (69592440 bytes), 65169 used (~128 by thread caches), needed_avg 66136, 0 failures, 2 users, @0x27064c0 [SHARED]
  - Pool fcgi_conn (1240 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x272ecc0 [SHARED]
  - Pool h2c (1304 bytes) : 22967 allocated (29948968 bytes), 22934 used (~30 by thread caches), needed_avg 22950, 0 failures, 1 users, @0x272ac00 [SHARED]
  - Pool quic_conn (4088 bytes) : 9659 allocated (39485992 bytes), 9512 used (~16 by thread caches), needed_avg 9564, 0 failures, 1 users, @0x2712700 [SHARED]
  - Pool hpack_tbl (4096 bytes) : 22965 allocated (94064640 bytes), 22930 used (~26 by thread caches), needed_avg 22933, 0 failures, 1 users, @0x2829440 [SHARED]
  - Pool quic_tx_rin (4120 bytes) : 0 allocated (0 bytes), 0 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x271a8c0 [SHARED]
  - Pool buffer (16384 bytes) : 18248 allocated (298975232 bytes), 17840 used (~62 by thread caches), needed_avg 18079, 0 failures, 1 users, @0x28234c0 [SHARED]
  - Pool trash (16416 bytes) : 8 allocated (131328 bytes), 8 used (~7 by thread caches), needed_avg 7, 0 failures, 1 users, @0x27471c0
  - Pool quic_conn_r (65560 bytes) : 9660 allocated (633309600 bytes), 9505 used (~9 by thread caches), needed_avg 9550, 0 failures, 2 users, @0x2716800 [SHARED]
  - Pool quic_rxbuf (262168 bytes) : 8 allocated (2097344 bytes), 8 used (~0 by thread caches), needed_avg 4, 0 failures, 1 users, @0x2726b40 [SHARED]
Total: 24 pools, 6220663624 bytes allocated, 6200790312 used (~2266952 by thread caches).
```

### Additional Information

CentOS Linux release 7.9.2009 (Core)

Linux 4.19.188-10.el7.x86_64 #1 SMP Wed Apr 28 09:54:38 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

**Comment by a-denoyelle** [comment]
Indeed this looks like a leak. First of all, can you update to the newly released 2.8.4 as there is some fixes related to memory leak. If this is not sufficient, you should run your process using `-dMno-merge` as by default pools are merged so it's not possible to identify where the leak occurs. Then you can rerun `show pools byusage`.

**Comment by chnchi** [comment]
> Indeed this looks like a leak. First of all, can you update to the newly released 2.8.4 as there is some fixes related to memory leak. If this is not sufficient, you should run your process using `-dMno-merge` as by default pools are merged so it's not possible to identify where the leak occurs. Then you can rerun `show pools byusage`.

Thank a lot for your suggestion. I have compiled version 2.8.4 of haproxy and replaced the previous version with it. I will observe its performance for a period of time and then provide feedback on how the new version is running. Thank you very much!

**Comment by wtarreau** [comment]
That makes me think that we should condider backporting 5714aff4a ("DEBUG: pool: store the memprof bin on alloc() and update it on free()") to the next 2.8, as it will show where leaked elements are allocated.

**Comment by chnchi** [comment]
After running version 2.8.4 of haproxy on one of the servers for some days, I have observed that the memory increase issue in version 2.8.4 is better than in version 2.8.3. 
However, there is still a memory leak problem in 2.8.4, as the memory usage continues to show an upward trend during running.
<img width="721" alt="image" src="https://github.com/haproxy/haproxy/assets/30403780/947f0a6b-68cd-4337-929f-de153fa45018">


Additionally, I’ve noticed that the total memory displayed by ‘show pool’ is significantly less than what is seen with ‘pidstat’ or ‘top’ command. Could it be that some memory usage is not being accounted for?

**Below is memory usage information:**

pidstat -p 10695 -r 1
05:36:31 PM   UID       PID  minflt/s  majflt/s     VSZ    RSS   %MEM  Command
05:36:32 PM  1002     10695      0.00      0.00 5512408 5159524  31.95  haproxy

<img width="535" alt="image" src="https://github.com/haproxy/haproxy/assets/30403780/704c6d3e-3a54-4ad6-bfa6-da67bf798f17">

10695> show pools byusage
Dumping pools usage. Use SIGQUIT to flush them.
  - Pool quic_rx_pac (376 bytes) : 3050625 allocated (1147035000 bytes), 3050625 used (~54 by thread caches), needed_avg 3050621, 0 failures, 1 users, @0x1da9800 [SHARED]
  - Pool hpack_tbl (4096 bytes) : 96573 allocated (395563008 bytes), 79263 used (~46 by thread caches), needed_avg 82680, 0 failures, 1 users, @0x1fab400 [SHARED]
  - Pool quic_conn_r (65560 bytes) : 4917 allocated (322358520 bytes), 3999 used (~17 by thread caches), needed_avg 3967, 0 failures, 1 users, @0x1dbdc00 [SHARED]
  - Pool buffer (16384 bytes) : 9329 allocated (152846336 bytes), 7580 used (~58 by thread caches), needed_avg 7527, 0 failures, 1 users, @0x1fa5480 [SHARED]
  - Pool h2c (1304 bytes) : 96573 allocated (125931192 bytes), 79259 used (~42 by thread caches), needed_avg 82724, 0 failures, 1 users, @0x1df6780 [SHARED]
  - Pool quic_crypto (1048 bytes) : 33232 allocated (34827136 bytes), 27532 used (~105 by thread caches), needed_avg 27162, 0 failures, 1 users, @0x1d9d580 [SHARED]
  - Pool quic_conn (4088 bytes) : 4954 allocated (20251952 bytes), 4005 used (~23 by thread caches), needed_avg 3981, 0 failures, 1 users, @0x1db5a80 [SHARED]
  - Pool connection (184 bytes) : 105809 allocated (19468856 bytes), 87927 used (~104 by thread caches), needed_avg 90815, 0 failures, 1 users, @0x1e3b540 [SHARED]
  - Pool task (152 bytes) : 113760 allocated (17291520 bytes), 95855 used (~57 by thread caches), needed_avg 99129, 0 failures, 1 users, @0x1e63d80 [SHARED]
  - Pool sockaddr (152 bytes) : 105823 allocated (16085096 bytes), 87925 used (~102 by thread caches), needed_avg 90751, 0 failures, 1 users, @0x1e333c0 [SHARED]
  - Pool session (152 bytes) : 105713 allocated (16068376 bytes), 87856 used (~38 by thread caches), needed_avg 91648, 0 failures, 1 users, @0x1e84400 [SHARED]
  - Pool ssl_sock_ct (152 bytes) : 101874 allocated (15484848 bytes), 83983 used (~27 by thread caches), needed_avg 87806, 0 failures, 1 users, @0x1d8d240 [SHARED]
  - Pool tasklet (56 bytes) : 206319 allocated (11553864 bytes), 171157 used (~109 by thread caches), needed_avg 176357, 0 failures, 1 users, @0x1e5fc80 [SHARED]
  - Pool quic_tx_pac (152 bytes) : 46200 allocated (7022400 bytes), 38915 used (~156 by thread caches), needed_avg 38879, 0 failures, 1 users, @0x1da5740 [SHARED]
  - Pool qcs (344 bytes) : 18161 allocated (6247384 bytes), 14806 used (~56 by thread caches), needed_avg 14715, 0 failures, 1 users, @0x1dc5dc0 [SHARED]
  - Pool stk_ctr (56 bytes) : 105725 allocated (5920600 bytes), 87875 used (~56 by thread caches), needed_avg 91479, 0 failures, 1 users, @0x1faf900 [SHARED]
  - Pool qc_stream_d (152 bytes) : 19366 allocated (2943632 bytes), 15908 used (~48 by thread caches), needed_avg 15785, 0 failures, 1 users, @0x1dea500 [SHARED]
  - Pool quic_tls_ke (56 bytes) : 42955 allocated (2405480 bytes), 35631 used (~125 by thread caches), needed_avg 35175, 0 failures, 1 users, @0x1dd6100 [SHARED]
  - Pool quic_rxbuf (262168 bytes) : 8 allocated (2097344 bytes), 8 used (~0 by thread caches), needed_avg 5, 0 failures, 1 users, @0x1de2380 [SHARED]
  - Pool qcc (376 bytes) : 4803 allocated (1805928 bytes), 3876 used (~15 by thread caches), needed_avg 3860, 0 failures, 1 users, @0x1dc9e80 [SHARED]
  - Pool quic_connec (152 bytes) : 9812 allocated (1491424 bytes), 7973 used (~35 by thread caches), needed_avg 7902, 0 failures, 1 users, @0x1db19c0 [SHARED]
  - Pool quic_tls_se (56 bytes) : 24047 allocated (1346632 bytes), 19824 used (~80 by thread caches), needed_avg 19620, 0 failures, 1 users, @0x1dde280 [SHARED]
  - Pool quic_cstrea (88 bytes) : 14649 allocated (1289112 bytes), 11992 used (~46 by thread caches), needed_avg 11908, 0 failures, 1 users, @0x1d994c0 [SHARED]
  - Pool quic_tls_iv (24 bytes) : 42957 allocated (1030968 bytes), 35632 used (~126 by thread caches), needed_avg 35175, 0 failures, 1 users, @0x1dda1c0 [SHARED]
  - Pool h3s (56 bytes) : 18166 allocated (1017296 bytes), 14791 used (~41 by thread caches), needed_avg 14721, 0 failures, 1 users, @0x1dcdf40 [SHARED]
  - Pool quic_arng (56 bytes) : 17752 allocated (994112 bytes), 14886 used (~50 by thread caches), needed_avg 14844, 0 failures, 1 users, @0x1d91340 [SHARED]
  - Pool quic_conn_c (152 bytes) : 4961 allocated (754072 bytes), 3998 used (~16 by thread caches), needed_avg 3982, 0 failures, 1 users, @0x1db9b40 [SHARED]
  - Pool h3c (120 bytes) : 4810 allocated (577200 bytes), 3878 used (~17 by thread caches), needed_avg 3860, 0 failures, 1 users, @0x1dd2040 [SHARED]
  - Pool trash (16416 bytes) : 17 allocated (279072 bytes), 17 used (~16 by thread caches), needed_avg 14, 0 failures, 1 users, @0x1eb8e80
  - Pool quic_frame (184 bytes) : 1348 allocated (248032 bytes), 1316 used (~183 by thread caches), needed_avg 1340, 0 failures, 1 users, @0x1d95400 [SHARED]
  - Pool quic_dgram (344 bytes) : 140 allocated (48160 bytes), 108 used (~102 by thread caches), needed_avg 117, 0 failures, 1 users, @0x1dad8c0 [SHARED]
  - Pool requri (1048 bytes) : 44 allocated (46112 bytes), 44 used (~44 by thread caches), needed_avg 39, 0 failures, 1 users, @0x1f3cc40 [SHARED]
  - Pool stream (888 bytes) : 46 allocated (40848 bytes), 46 used (~45 by thread caches), needed_avg 39, 0 failures, 1 users, @0x1e12d40 [SHARED]
  - Pool http_txn (216 bytes) : 55 allocated (11880 bytes), 55 used (~55 by thread caches), needed_avg 45, 0 failures, 1 users, @0x1e16e00 [SHARED]
  - Pool h2s (184 bytes) : 50 allocated (9200 bytes), 50 used (~50 by thread caches), needed_avg 42, 0 failures, 1 users, @0x1df26c0 [SHARED]
  - Pool h1s (184 bytes) : 50 allocated (9200 bytes), 50 used (~50 by thread caches), needed_avg 43, 0 failures, 1 users, @0x1e02a00 [SHARED]
  - Pool qc_stream_b (56 bytes) : 130 allocated (7280 bytes), 108 used (~35 by thread caches), needed_avg 122, 0 failures, 1 users, @0x1de6440 [SHARED]
  - Pool stconn (88 bytes) : 66 allocated (5808 bytes), 66 used (~64 by thread caches), needed_avg 57, 0 failures, 1 users, @0x1e477c0 [SHARED]
  - Pool sedesc (56 bytes) : 65 allocated (3640 bytes), 60 used (~54 by thread caches), needed_avg 54, 0 failures, 1 users, @0x1e43700 [SHARED]
  - Pool h1c (184 bytes) : 11 allocated (2024 bytes), 5 used (~0 by thread caches), needed_avg 9, 0 failures, 1 users, @0x1e06ac0 [SHARED]
  - Pool conn_hash_n (56 bytes) : 10 allocated (560 bytes), 6 used (~1 by thread caches), needed_avg 9, 0 failures, 1 users, @0x1e37480 [SHARED]
  - Pool sig_handler (56 bytes) : 7 allocated (392 bytes), 7 used (~0 by thread caches), needed_avg 4, 0 failures, 1 users, @0x1eb0cc0 [SHARED]
  - Pool appctx (280 bytes) : 1 allocated (280 bytes), 1 used (~0 by thread caches), needed_avg 0, 0 failures, 1 users, @0x1e90640 [SHARED]
  - Pool mux_pt (56 bytes) : 2 allocated (112 bytes), 2 used (~1 by thread caches), needed_avg 1, 0 failures, 1 users, @0x1e7c240 [SHARED]
Total: 44 pools, 2332421888 bytes allocated, 2118016320 used (~3142616 by thread caches).

**Comment by a-denoyelle** [comment]
It seems the issue comes from `quic_rx_packet` pool. A recent fix related to it was done, alas it was too late and not included in 2.8.4. By any chance, are you able to recompile haproxy to test it ? The simple way would be to use the 2.9-dev11. If you are unsure to test a development release, you could apply the following patch on top of the 2.8.4.

[0001-BUG-MINOR-quic-Possible-RX-packet-memory-leak-under-.patch.txt](https://github.com/haproxy/haproxy/files/13486587/0001-BUG-MINOR-quic-Possible-RX-packet-memory-leak-under-.patch.txt)

**Comment by chnchi** [comment]
> It seems the issue comes from `quic_rx_packet` pool. A recent fix related to it was done, alas it was too late and not included in 2.8.4. By any chance, are you able to recompile haproxy to test it ? The simple way would be to use the 2.9-dev11. If you are unsure to test a development release, you could apply the following patch on top of the 2.8.4.
> 
> [0001-BUG-MINOR-quic-Possible-RX-packet-memory-leak-under-.patch.txt](https://github.com/haproxy/haproxy/files/13486587/0001-BUG-MINOR-quic-Possible-RX-packet-memory-leak-under-.patch.txt)

Thank you for your response. I have recompiled version 2.9-dev11 and replaced it on the server. I will wait and observe for a while before providing feedback on the performance of the new version. Thanks a lot!

**Comment by a-denoyelle** [comment]
Thank you for your efforts, hope this will be sufficient to fix the situation.

**Comment by a-denoyelle** [comment]
FYI, we have just found a bug with HTTP/3 which can cause pages of your websites to appear as broken without any explicit error. The issue lies on the H3 zero-copy forwarding which was just enable for 2.9-dev11.  To prevent it from occurring and disturbing your analysis on your own issue, please use the following statement on the global section `tune.disable-zero-copy-forwarding`. Sorry for the inconvenience.

**Comment by chnchi** [comment]
Thank you for the reminder. I will adjust the configuration according to your suggestions later on.

**Comment by wtarreau** [comment]
@a-denoyelle do you have an idea for a fix already or should we temporarily revert (or disable) the feature for dev12 ?

**Comment by a-denoyelle** [comment]
@wtarreau capflam has already a fix which I am currently testing.

**Comment by a-denoyelle** [comment]
@chnchi in fact, after I check carefully, `tune.disable-zero-copy-forwarding` will not fix correctly the issue. We are currently testing a patch which I can provide you soon enough. Again, very sorry for this mess.

**Comment by chnchi** [comment]
> @chnchi in fact, after I check carefully, `tune.disable-zero-copy-forwarding` will not fix correctly the issue. We are currently testing a patch which I can provide you soon enough. Again, very sorry for this mess.

Message received. So far, I haven’t noticed any impact on the business. Once again, thank you for your work!

**Comment by a-denoyelle** [comment]
@chnchi the fix has just been pushed. I think the simplest solution is to switch directly to the master HEAD of the git repository which also has the memleak patch integrated.

**Comment by chnchi** [comment]
Okay, I will recompile using the master branch and then deploy it for testing.

**Comment by chnchi** [comment]
The 2.9-dev version has been deployed on the server and running for two days. Monitoring that memory usage remains stable. It seems that the memory leak issue after enabling QUIC has been fixed. Thank you all for your help!

<img width="1638" alt="image" src="https://github.com/haproxy/haproxy/assets/30403780/e3d4eccd-20c7-4b42-bcdd-46a70266150b">

**Comment by a-denoyelle** [comment]
Great, thanks for your feedback !

**Comment by wtarreau** [comment]
OK marking as fixed, still needs to be backported to 2.8.

**Comment by wtarreau** [comment]
Reopening, not yet backported!
